sklearn/cluster/hierarchical.py

```diff
@@ -444,6 +447,39 @@ def linkage_tree(X, connectivity=None, n_components=None,
         assert n_clusters <= n_samples
         n_nodes = 2 * n_samples - n_clusters
 
+    if linkage == 'single':
+        from scipy.sparse.csgraph import minimum_spanning_tree
+
+        # Use scipy.sparse.csgraph to generate a minimum spanning tree
+        mst = minimum_spanning_tree(connectivity.tocsr())
+
+        # Convert the graph to scipy cluster array format
+        nonzeros = mst.nonzero()
+        nonzero_vals = mst.data
+        mst_array = np.vstack(nonzeros + (nonzero_vals,)).T
+
+        # Sort edges of the min_spanning_tree by weight
+        mst_array = mst_array[np.argsort(mst_array.T[2]),:]
+
+        # Convert edge list into standard hierarchical clustering format
+        single_linkage_tree = _hierarchical.single_linkage_label(mst_array)
+        children_ = single_linkage_tree[:, :2].astype(np.int)
+
+        # Compute parents
+        parent = np.arange(n_nodes, dtype=np.intp)
+        for i, (left, right) in enumerate(children_, n_samples):
```

Comment:
Isn't this slow?

Reply:
sorry can you explain the difference between the two graphs again? The cythonization of the tree creation?

Reply:
Both graphs use the same code in hierarchy_.py, the difference is in the data passed to in. In the first graph the sparse matrix that is passed in is not guaranteed to have a single connected component, which then requires some pre-processing to "fix" that by adding extra entries to the matrix. This pre-processing is common to all the linkage approaches, but isn't part of the algorithms per se (but the fit method will do it if it is required). The second graph simple ensures that the required pre-processing is done *before* the fit method is called, and the timing is thus only for the actual linkage methods, and no longer includes the common pre-processing step (which can be time consuming).